lib/inventory/inventory.go

```diff
@@ -260,30 +272,138 @@ type UpstreamHandle interface {
 	// Hello gets the cached upstream hello that was used to initialize the stream.
 	Hello() proto.UpstreamInventoryHello
 
-	Ping(ctx context.Context) (d time.Duration, err error)
+	Ping(ctx context.Context, id uint64) (d time.Duration, err error)
 	// HasService is a helper for checking if a given service is associated with this
 	// stream.
 	HasService(types.SystemRole) bool
+
+	// StateTracker gets the instance state tracker.
+	//
+	// NOTE: The returnd state tracker is *not* locked, and must be locked prior to any
+	// access. Prefer using the InstanceStateTracker.WithLock method where possible.
+	StateTracker() *InstanceStateTracker
+
+	// HeartbeatInstance triggers an early instance heartbeat. This function does not
+	// wait for the instance heartbeat to actually be completed, so calling this and then
+	// immediately locking the InstanceStateTracker will likely result in observing the
+	// pre-heartbeat state.
+	HeartbeatInstance()
+}
+
+// InstanceStateTracker tracks the state of a connected instance from the point of view of
+// the inventory controller. Values in this struct tend to be lazily updated/consumed. For
+// example, the LastHeartbeat value is nil until the first attempt to heartbeat the instance
+// has been made *for this TCP connection*, meaning that you can't distinguish between an instance
+// that just joined the cluster from one that just reconnected by looking at this struct. Similarly,
+// when using this struct to inject control log entries, said entries won't be included until the next
+// normal instance heartbeat (though this may be triggered early via UpstreamHandle.HeartbeatInstance()).
+// The primary intended usage pattern is for periodic operations to append entries to the QualifiedPendingControlLog,
+// and then observe wether or not said entries end up being included on subsequent iterations. This
+// patterns lets us achieve a kind of lazy "locking", whereby complex coordination can occur without
+// large spikes in backend load. See the QualifiedPendingControlLog field for an example of this pattern.
+type InstanceStateTracker struct {
```

Comment:
`handle.RTFM_GetStateTracker()`?

In seriousness, I've considered a few alternatives.  The best alternative I could come up with was a closure-style interaction that passes in previous state + pending state and lets you return additional pending state.  Ex:

```golang
type PendingControlLogEntries struct {
    Qualified []types.InstanceControlLogEntry
    Unqualified []types.InstanceControlLogEntry
}

handle.VisitControlLog(func(lastHeartbeat types.Instance, pending PendingControlLogEntries) PendingControlLogEntries {
    var additions PendingControlLogEntries
    if condition(lastHeartbeat, pending) {
        additions.Qualified = entriesToAdd
    }
    return additions
})
```

This could do a defensive copy to ensure that callers can't break stuff by modifying the inputs, making it a "pure" function.

Ended up going with the flatter tracker because I felt that this strategy was starting to add too much complexity to the handle/tracker code, but it might be worth it for the reduction in footgun.

Thoughts/alternatives?